llvm/lib/SYCLLowerIR/SYCLConditionalCallOnDevice.cpp

```diff
@@ -0,0 +1,120 @@
+//===-- SYCLConditionalCallOnDevice.cpp - SYCLConditionalCallOnDevice Pass
+//--===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Pass performs transformations on functions which represent the conditional
+// call to application's callable object. The conditional call is based on the
+// SYCL device's aspects or architecture passed to the functions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/SYCLLowerIR/SYCLConditionalCallOnDevice.h"
+
+#include "llvm/IR/Function.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+
+using namespace llvm;
+
+cl::opt<std::string>
+    UniquePrefixOpt("sycl-conditional-call-on-device-unique-prefix",
+                    cl::Optional, cl::Hidden,
+                    cl::desc("Set unique prefix for a translation unit, "
+                             "required for funtions with external linkage"),
+                    cl::init(""));
+
+PreservedAnalyses
+SYCLConditionalCallOnDevicePass::run(Module &M, ModuleAnalysisManager &) {
+  // find call_if_on_device_conditionally function
+  SmallVector<Function *, 4> FCallers;
+  for (Function &F : M.functions()) {
+    if (F.isDeclaration())
+      continue;
+
+    if (CallingConv::SPIR_KERNEL == F.getCallingConv())
+      continue;
+
+    if (F.getName().contains("call_if_on_device_conditionally") &&
+        !F.getName().contains("call_if_on_device_conditionally_helper"))
+      FCallers.push_back(&F);
+  }
+
+  int FCallerIndex = 1;
+  for (Function *FCaller : FCallers) {
+    // Find call to @CallableXXX in call_if_on_device_conditionally function
```

Comment:
Thanks for this comment. Note that even though C++ source code may only contain a single call, corresponding LLVM IR may contain several `call` instructions within a function: those could be calls to various intrinsics. I recommend to run the pass on a real-life example compiled with `-O0` and/or `-g` and one of such runs (`-O0` probably) should be added as a test. You probably need an early exit from instructions loop below if `CallInst` is an intrinsic call (perhaps could be detected by checking [`getIntrinsicID()`](https://llvm.org/docs/doxygen/classllvm_1_1CallBase.html#ac62778065b99372cc62cf994b967e7e8) result)

Reply:
Debug LLVM IR has only one call, no intrinsic calls, and I added a check for `getIntrinsicID()` just in case.
Will it be good to throw some static assert in case the function has more than one non-intrinsic call?

Reply:
> Will it be good to throw some static assert in case the function has more than one non-intrinsic call?

You can't do `static_assert` like this, but if we are sure about a single call, then `assert` won't hurt